### ISSUE TITLE
move_and_slide() changes x and z trajectory when climbing slopes that count as floors

### ISSUE THREAD

**Issue by Gamepro5** [godotengine/godot#63177]
### Godot version

4.0 alpha 12

### System information

Windows 11, Vulkan, Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz   2.30 GHz

### Issue description

Moving up a slope that is within the `floor_max_angle` using `move_and_slide()` behaves unlike any primitive movement system in other game engines. Instead of simply adjusting the y velocity to keep up the slope, it slightly adjusts the x and z velocities. 

The expected behavior would be that climbing a slope that counts as the "floor" would have zero impact on your x and z trajectory, that it would simply adjust your Y velocity to climb the slope. This behavior is expected because this behavior is what the basic character controller does in both Unity and Unreal Engine, as well as in the Source engine. (Replicate steps 3 to 7 on these engines to see)

I've been trying to script a workaround for this since Godot 3. The closest I've gotten to fixing this was [this](https://github.com/Gamepro5/Custom_Godot_3D_KinematicBody_Controller/blob/main/KinematicBody.gd), but there were some collision detection issues as it was still using the bullet physics engine. This issues seems to be too deeply engrained into how `move_and_slide()` for me to find a workaround.

### Steps to reproduce

1. Open the project
2. Click Play
3. Climb the slope, position yourself such that your center of the screen (indicated with the visible collision boxes option) is lined up with the top corner of any slope.
4. Let go of your mouse, do not change your rotation
5. Press S to go down the slope
6. Press W to go up the slope again
7. Observe as you are not lined up with the corner as you once were

### Minimal reproduction project

[Example.zip](https://github.com/godotengine/godot/files/9136925/Example.zip)

**Comment by L1ghtingBolt** [comment]
Does this happen in normal Godot?

**Comment by Calinou** [comment]
I can confirm this on 4.0.alpha 5fec0d232.

To clarify, what happens is that the character does work in a logical fashion when walking up/down slopes, but your position is offset over time when walking down/up a slope. Adjusting physics ticks per second does not affect this issue (if anything, it appears to make it slightly worse). This may be due to floating-point precision issues.

### Physics Ticks Per Second = 60 (default)

https://user-images.githubusercontent.com/180032/179809989-8ae24f47-f2d5-4cd1-92a5-ad549aaa9787.mp4

https://user-images.githubusercontent.com/180032/179809998-d8f84bef-20a6-4152-a24e-906236c64af0.mp4

### Physics Ticks Per Second = 240

https://user-images.githubusercontent.com/180032/179809980-22d07e18-e744-4e51-a55a-a4238bb8d7cd.mp4


https://user-images.githubusercontent.com/180032/179809973-4d7ddee5-1ffb-4214-851b-31973ab205ad.mp4

**Comment by Gamepro5** [comment]
> This isn't always desired as you may want the player to slow down when climbing a slope (without preventing them from climbing the slope entirely). This is especially the case in a game with more realistic movement.

I agree, but adding this functionality for the user in GDScript would be much easier than trying to wrestle the physics engine to try to get it to behave like how most games do slope movement, hence why I think it should be fixed.

**Comment by wareya** [comment]
The way that horizontal velocity and direction is generally retained by these games (I know this is true of Source engine and early id tech games, at the very least) is by using a "stair stepping" algorithm even if the floor is a ramp instead of a staircase. "Stair stepping" algorithms generally move the character up, then sideways, then down, and make sure that the resulting set of motions look like what you would expect from stepping up onto higher ground. The actor does not actually travel up the slope, and as such, does not need to make its movement vector "slide" along its collision normal. The "sliding" step is what causes the horizontal velocity and direction to change, and not doing it avoids this issue.

You can see an example of such logic here (creative commons zero license): <https://github.com/wareya/QuakelikeBody/blob/0a012b757a70dfcc2fc4f03b65a7f15e2dc1a52a/QuakelikeBody.gd#L151> <https://github.com/wareya/QuakelikeBody/blob/0a012b757a70dfcc2fc4f03b65a7f15e2dc1a52a/QuakelikeBody.gd#L226>

"Stair stepping" logic usually doesn't work that well with character collision hulls that have steep slope-like parts, like capsules, and it can also be glitchy if the collision system handles collision margins in inconsistent or physically incoherent ways. In such cases, you may need to detect the floor-like-ness of a collision surface through means other than checking the collision normal, like by associating specific triangles/edges/materials with "is a wall" or "is a floor" flags, or change your collision hull to a shape that doesn't have steep-but-not-horizontal sides, or fix the margin handling bugs in your collision system. (OpenMW had such a bug back when it used capsules instead of AABBs for collision, and fixed it by switching characters to AABBs: <https://gitlab.com/OpenMW/openmw/-/issues/4247>)

**Comment by Gamepro5** [comment]
This issue led me to make my own character controller from scratch using only `move_and_collide()`. The piece of code responsible for the "no impact on x/z trajectory" in my overhaul is as follows, perhaps it should be added to the official `move_and_slide()` function. It uses the dot product formula re-written to solve for just the y velocity.

```python
if floor_collision_normal != Vector3.ZERO:
	on_floor = true
	vel.y = 0;
	vel.y = (-floor_collision_normal.z*vel.z-floor_collision_normal.x*vel.x)/floor_collision_normal.y
	if (floor_collision_normal.y == 0): #safeguard. if the y normal of the slope is 0, it means you are trying to climb a completely vertical wall. This is a divided by zero error, which evaluates to infinity. This is invalid. In other words, there is no velocity you can give to the y component to climb a vertical wall without changing the x and z velocity.
		vel.y = 0
```

My overhaul can be found [here](https://github.com/Gamepro5/Custom_CharacterBody3D_Movement_System/blob/main/player/CharactedController.gd).

**Comment by Calinou** [comment]
https://github.com/Calinou/godot-demo-projects/tree/add-first-person-shooter uses a custom implementation of `is_on_floor()` using a ShapeCast3D to provide reliable stair stepping. You may be interested in that :slightly_smiling_face:

**Comment by Zireael07** [comment]
That's amazing Calinou, I hope it'll get added to the demo projects repo soon!
